### PR TITLE
feature: add a refactored copy of botticelli

### PR DESCRIPTION
**Pull request by bassosimone**
This diff is a quick refactor of [neubot/botticelli](
https://github.com/neubot/botticelli) that compiles and
will hopefully work. I started moving around all the
pieces that were part of botticelli to ensure that the
overall code is more tidy.

---
This change is [Reviewable](https://reviewable.io/reviews/m-lab/ndt-server-go/15)

### CODE REVIEW

**Comment by gfr10598** [comment]
*[ndt-server.go, line 1 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15#-L3Yp_NpGHNP3sJEk-VX:-L3Yp_NpGHNP3sJEk-VY:br8151a) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/ndt-server.go#L1)):*
> ```Go
> package main
> ```

This should go under cmd/ndt-server/ndt-server.go

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15)*

**Comment by gfr10598** [comment]
Review status: 0 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[protocol/messages.go, line 26 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15#-L3YrhoUgKq6Z94iWCx_:-L3YrhoUgKq6Z94iWCxa:b-l5tghj) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/protocol/messages.go#L26)):*
> ```Go
> */
> 
> func read_message_internal(cc net.Conn, reader io.Reader) (
> ```

readMessageInternal

---

*[util/io.go, line 1 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15#-L3Yq6yGkJ1ES2fMj9uq:-L3Yq6yH3UBP4VK5eVsn:b3b43xa) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L1)):*
> ```Go
> // Part of Neubot <https://neubot.nexacenter.org/>.
> ```

Let's pull util into a separate PR, and add unit tests.

---

*[util/io.go, line 17 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15#-L3YqwoOyelUxYRN2Yco:-L3YqwoOyelUxYRN2Ycp:b-fhffdy) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L17)):*
> ```Go
> 	"time"
> )
> ```

type DeadlineReader struct {
   io.Reader   // this inherits all methods of reader.  We will override them
   conn net.Conn  // This encapsulates the conn for deadlines.
   deadline time.Duration
}

func NewDeadlineReader(conn net.Conn) {
   // initialize deadline constant
   // create reader
   // create and return new struct
}

func (conn DeadlineReader) ReadFull(data) error {
   ...
}

---

*[util/io.go, line 30 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15#-L3YqT4H8bjdLVDzzw8c:-L3YqT4H8bjdLVDzzw8d:bbe0mwr) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L30)):*
> ```Go
> 		return count, err
> 	}
> 	err = conn.SetReadDeadline(time.Time{})
> ```

return count, conn.SetReadDeadline(time.Time{})

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15)*

**Comment by gfr10598** [comment]
Review status: 0 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[util/io.go, line 1 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15#-L3Yq6yGkJ1ES2fMj9uq:-L3YtjXa3H08pwZ5CXeF:bm3hwtz) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L1)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Let's pull util into a separate PR, and add unit tests.
</blockquote></details>

Actually, let's start with just util/random first.

---

*[util/io.go, line 30 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15#-L3YqT4H8bjdLVDzzw8c:-L3YtRz1GdfJG9FlH-i4:b-f0ou9s) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L30)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

return count, conn.SetReadDeadline(time.Time{})
</blockquote></details>

(If we keep this outline)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15)*

**Comment by gfr10598** [comment]
Review status: 0 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[util/io.go, line 1 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15#-L3Yq6yGkJ1ES2fMj9uq:-L3Yu4IU4-GkK87lkIdQ:batuklf) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L1)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Actually, let's start with just util/random first.
</blockquote></details>

For this one, maybe use package ioutil

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15)*

**Comment by bassosimone** [comment]
Started moving the code over in #18

File: protocol/constants.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package protocol
+
+// Message types:
+
+const KvCommFailure byte = 0
```
Comment (bassosimone) [inline protocol/constants.go]: TODO(bassosimone): use `iota` to initialise variables

File: protocol/constants.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package protocol
```
Comment (bassosimone) [inline protocol/constants.go]: This is part of the protocol

File: protocol/constants.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+// Message types:
+
+const KvCommFailure byte = 0
+const KvSrvQueue byte = 1
+const KvMsgLogin byte = 2
+const KvTestPrepare byte = 3
+const KvTestStart byte = 4
```
Comment (bassosimone) [inline protocol/constants.go]: Drop the `kv` prefix

File: protocol/constants.go (hunk 2 above)
Comment (bassosimone) [reply protocol/constants.go]: It can also be at the top of one file

File: protocol/control.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+			break
+		}
+		kv_test_pending_mutex.Unlock()
+		err = update_queue_pos(rdwr, 1)
+		if err != nil {
+			log.Println("ndt: failed to update client of its queue position")
+			return
+		}
+		time.Sleep(3.0 * time.Second)
+	}
```
Comment (bassosimone) [inline protocol/control.go]: TODO: remove the queue before submitting PR since it's suboptimal

File: protocol/control.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+		log.Println("ndt: cannot write kickoff message")
+		return
+	}
+
+	// Queue management
+	// XXX The current implementation of queue management is minimal, and
+	// possibly also very ugly and stupid. Must be improved.
+	//
+	// Moreover the lock/unlock dance with the mutex is not idiomatic
+	// golang and it would be better to use messages and channels.
```
Comment (bassosimone) [inline protocol/control.go]: TODO: separate before the queue and after the queue

File: protocol/meta.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
+package protocol
+
```
Comment (bassosimone) [inline protocol/meta.go]: single package for all the tests; let's use `nettests`

File: protocol/s2c.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+// RunS2cTest runs the S2C test. |rdwr| is the buffered reader and writer
+// for the connection. |is_extended| is true when we want to run a multi-stream
+// test. Returns the error.
+func RunS2cTest(rdwr *bufio.ReadWriter, is_extended bool) error {
+
+	// Bind port and tell the port number to the server
+	// TODO: choose a random port instead than an hardcoded port
+
+	listener, err := net.Listen("tcp", ":3010")
```
Comment (bassosimone) [inline protocol/s2c.go]: Let's take code for random port from the "start" branch!

File: protocol/s2c.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+/*
+ ____ ____   ____
+/ ___|___ \ / ___|
+\___ \ __) | |
+ ___) / __/| |___
+|____/_____|\____|
+
+*/
+
+const kv_parallel_streams int = 3
```
Comment (bassosimone) [inline protocol/s2c.go]: Add reference to the original spec

**Comment by gfr10598** [comment]
Review status: 0 of 5 files reviewed at latest revision, 13 unresolved discussions.

---

*[nettests/meta.go, line 6 at r3](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-LD85CJbC4dCDOgH5Hbc:-LD85CJbC4dCDOgH5Hbd:b-qa067a) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b08e404db670a2413fe19646f115bf96a70ed8b5/nettests/meta.go#L6)):*
> ```Go
> 	"bufio"
> 	"errors"
>   "github.com/m-lab/ndt-server-go/protocol"
> ```

go fmt

---

*[nettests/meta.go, line 26 at r3](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-LD85pEr0_GJsdnGR1Bl:-LD85pEr0_GJsdnGR1Bm:bb69l0f) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b08e404db670a2413fe19646f115bf96a70ed8b5/nettests/meta.go#L26)):*
> ```Go
> 
> 	// Read a sequence of TEST_MSGs from client
> ```

extra newline

---


*Comments from [Reviewable](https://reviewable.io/reviews/m-lab/ndt-server-go/15)*

**Comment by bassosimone** [comment]
Review status: 0 of 5 files reviewed at latest revision, 15 unresolved discussions.

---

*[protocol/constants.go, line 1 at r2](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YpB9d4mlAB87sTL9I-r2-1:-LD8S-N7Dbl0OQeEdEK9:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e916e1b318eb754d27bc56c9d9f191ba7c127afb/protocol/constants.go#L1)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

It can also be at the top of one file
</blockquote></details>

Done.

---

*[protocol/constants.go, line 5 at r2](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YpB9d4mlAB87sTL9I-r2-5:-LD8RzS0E3ykmamKzwxj:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e916e1b318eb754d27bc56c9d9f191ba7c127afb/protocol/constants.go#L5)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

TODO(bassosimone): use `iota` to initialise variables
</blockquote></details>

Done.

---

*[protocol/constants.go, line 9 at r2](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YpB9d4mlAB87sTL9I-r2-9:-LD8S2_E7n80DHHxf3RJ:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e916e1b318eb754d27bc56c9d9f191ba7c127afb/protocol/constants.go#L9)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

Drop the `kv` prefix
</blockquote></details>

Done.

---

*[protocol/control.go, line 70 at r2](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YpB9d4mlAB87sTL9J-r2-70:-LD8S4js0pK07Ot6nKV5:bmg78ov) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e916e1b318eb754d27bc56c9d9f191ba7c127afb/protocol/control.go#L70)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

TODO: separate before the queue and after the queue
</blockquote></details>

Not relevant since we've decided not to have a queue

---

*[protocol/control.go, line 86 at r2](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YpB9d4mlAB87sTL9J-r2-86:-LD8S3ueBwjh1rEwyW29:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e916e1b318eb754d27bc56c9d9f191ba7c127afb/protocol/control.go#L86)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

TODO: remove the queue before submitting PR since it's suboptimal
</blockquote></details>

Done.

---

*[protocol/messages.go, line 26 at r1](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YrhoUgKq6Z94iWCx_:-LD8SCzmA7R6ohD_XNac:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/protocol/messages.go#L26)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

readMessageInternal
</blockquote></details>

Done.

---

*[protocol/meta.go, line 2 at r2](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YpB9d4mlAB87sTL9L-r2-2:-LD8R-rU2I4IdGO9igBy:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e916e1b318eb754d27bc56c9d9f191ba7c127afb/protocol/meta.go#L2)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

single package for all the tests; let's use `nettests`
</blockquote></details>

Done.

---

*[protocol/s2c.go, line 24 at r2](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YpB9d4mlAB87sTL9M-r2-24:-LD8ROsCCd5rqaAewriZ:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e916e1b318eb754d27bc56c9d9f191ba7c127afb/protocol/s2c.go#L24)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

Add reference to the original spec
</blockquote></details>

Done.

---

*[protocol/s2c.go, line 40 at r2](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YpB9d4mlAB87sTL9M-r2-40:-LD8RrUD7LtPl0dudcZl:b3m4z5u) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e916e1b318eb754d27bc56c9d9f191ba7c127afb/protocol/s2c.go#L40)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

Let's take code for random port from the "start" branch!
</blockquote></details>

Used more or less equivalent code.

---

*[util/io.go, line 1 at r1](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3Yq6yGkJ1ES2fMj9uq:-LD8SH5G5dHXNzn5PHqK:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L1)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

For this one, maybe use package ioutil
</blockquote></details>

Done.

---

*[util/io.go, line 17 at r1](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YqwoOyelUxYRN2Yco:-LD8SJVC8NUw6_OKlQo4:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L17)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

type DeadlineReader struct {
   io.Reader   // this inherits all methods of reader.  We will override them
   conn net.Conn  // This encapsulates the conn for deadlines.
   deadline time.Duration
}

func NewDeadlineReader(conn net.Conn) {
   // initialize deadline constant
   // create reader
   // create and return new struct
}

func (conn DeadlineReader) ReadFull(data) error {
   ...
}
</blockquote></details>

Done.

---

*[util/io.go, line 30 at r1](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YqT4H8bjdLVDzzw8c:-LD8SIXq858XuWSuOc_4:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L30)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

(If we keep this outline)
</blockquote></details>

Done.

---

*[nettests/meta.go, line 6 at r3](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-LD85CJbC4dCDOgH5Hbc:-LD8R9Nd1cqGGKBnfdgq:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b08e404db670a2413fe19646f115bf96a70ed8b5/nettests/meta.go#L6)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

go fmt
</blockquote></details>

Done.

---

*[nettests/meta.go, line 26 at r3](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-LD85pEr0_GJsdnGR1Bl:-LD8RHeUBEoJ_rkKYyPT:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b08e404db670a2413fe19646f115bf96a70ed8b5/nettests/meta.go#L26)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

extra newline
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/m-lab/ndt-server-go/15)*

**Comment by bassosimone** [comment]
Review status: 0 of 5 files reviewed at latest revision, 15 unresolved discussions.

---

*[ndt-server.go, line 1 at r1](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3Yp_NpGHNP3sJEk-VX:-LD8SO756nRt-euJ_Eoy:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/ndt-server.go#L1)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

This should go under cmd/ndt-server/ndt-server.go
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/m-lab/ndt-server-go/15)*

File: control/control.go [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+		log.Println("ndt: cannot write kickoff message")
+		return
+	}
+	err = rdwr.Writer.Flush()
+	if err != nil {
+		log.Println("ndt: cannot flush kickoff message")
+		return
+	}
+
+	// Queue management (simplified)
```
Comment (bassosimone) [inline control/control.go:47]: May want to call this dispatcher rather than queue

**Comment by pboothe** [comment]
Closing this.

**Pull request closed without merging**